vmm/src/migration/state.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::result;
+use std::sync::{Arc, Mutex, RwLock};
+use std::collections::HashMap;
+use std::sync::mpsc::{channel, Sender, SendError, RecvError};
+use crate::api::{ApiResponsePayload, ApiResponse};
+
+/// MigrationState errors are sent back from the receiver through the ApiResponse.
+#[derive(Debug)]
+pub enum MigrationStateError {
+    /// API request receive error
+    ApiRequestRecv(RecvError),
+
+    /// API response send error
+    ApiResponseSend(SendError<ApiResponse>),
+
+    /// Cannot handle migration state load request.
+    MigrationStateLoad,
+}
+pub type MigrationStateResult<T> = result::Result<T, MigrationStateError>;
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct MigrationStateData {
+    pub state: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub enum MigrationResponsePayload {
+    /// No data is sent on the channel.
+    Empty,
+
+    /// Migration state content
+    MigrationState(MigrationStateData),
+}
+
+/// This is the response sent by the VMM API server through the mpsc channel.
+pub type MigrationStateResponse = std::result::Result<MigrationResponsePayload, MigrationStateError>;
+
+#[derive(Clone)]
+pub struct MigrationState {
```

Comment:
I want to do this in next phase, at "TODO" place, i.e. the patch to implement MigrationData. I want to save the states into MigrationData buffer.

Reply:
In that case, `MigrationState` is only a map of components, it does not hold any state. I think it would make sense to use `MigrationState` to hold both components and their states.

Reply:
Question is what it is used for? The states data is used to be saved into a file for snapshot, or in MigrationDataFile buffer to be transferred to target VM for live migration. So, I do not think it is necessary to keep the data in "MigrationState". How do you think?

Reply:
What I'm saying is that `MigrationDataFile` could be replaced with `MigrationState` because `MigrationDataFile` is essentially going to be a hashmap of components (key) and states (value). So instead of having 2 separate structures, I'm suggesting having one that would map components to both their mpsc Sender channel and their state.

Reply:
Ok, if we think ram data is state too, we can do such change. Per current thought, it sounds reasonable.